### ISSUE TITLE
`checkAnnotationEquality` crashes when comparing two lower-bounded wildcards with parameterized bounds

### ISSUE THREAD

**Issue by smillst** [typetools/checker-framework#8167]
Comparing the qualifiers of two equal proper types crashes with a `BugInCF` when both types are
lower-bounded wildcards whose bound is a parameterized type.  Inference should report a qualifier
conflict here; instead it crashes inside the check that would have reported it.

This is specific to the `proper-type-equality-qualifiers` branch.  `ProperType.checkAnnotationEquality`
does not exist on `master`, where `reduceEquality` returns `TRUE` for two proper types without
comparing qualifiers, so `isSubtype` is never reached on this path.

## Reproduce

Tainting Checker, Checker Framework 4.2.4-SNAPSHOT at commit `7a1017c`, JDK 21.

```java
import org.checkerframework.checker.tainting.qual.Tainted;
import org.checkerframework.checker.tainting.qual.Untainted;

public class LowerBoundedWildcardCrash {
  interface Sup<T> {}

  interface A<T> extends Sup<T> {}

  interface B<T> extends Sup<T> {}

  interface Pair<X, Y> {}

  interface Holder<X> {}

  <X, S extends A<Pair<X, ? super @Untainted Holder<X>>>> S m(X x) {
    throw new RuntimeException();
  }

  void use(String s) {
    B<Pair<String, ? super @Tainted Holder<String>>> x = m(s);
  }
}
```

```
javac -processor tainting -AconvertTypeArgInferenceCrashToWarning=false LowerBoundedWildcardCrash.java
```

## Expected

`type.arguments.not.inferred`.  The two upper bounds on `S` imply
`? super @Untainted Holder<X>` = `? super @Tainted Holder<String>`, whose lower bounds differ in
their qualifiers, so no instantiation of `S` exists.

## Actual

```
error: AsSuperVisitor: type is not an erased subtype of supertype.
  type: @Tainted Object
  superType: @Tainted Holder<@Untainted String>
  ; The Checker Framework crashed.

  at AsSuperVisitor.errorTypeNotErasedSubtypeOfSuperType(AsSuperVisitor.java:148)
  at AsSuperVisitor.visitDeclared_Declared(AsSuperVisitor.java:354)
  at AnnotatedTypes.castedAsSuper(AnnotatedTypes.java:154)
  at DefaultTypeHierarchy.visitTypeArgs(DefaultTypeHierarchy.java:634)
  at DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:602)
  at DefaultTypeHierarchy.visitType_Wildcard(DefaultTypeHierarchy.java:1314)
  at DefaultTypeHierarchy.visitDeclared_Wildcard(DefaultTypeHierarchy.java:765)
  at DefaultTypeHierarchy.visitWildcard_Type(DefaultTypeHierarchy.java:1354)
  at DefaultTypeHierarchy.visitWildcard_Wildcard(DefaultTypeHierarchy.java:1148)
  at DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:154)
  at ProperType.checkAnnotationEquality(ProperType.java:217)
  at Typing.reduceEquality(Typing.java:473)
```

With `-AconvertTypeArgInferenceCrashToWarning` at its default of `true`, this surfaces as
`error: [type.argument.inference.crashed]` rather than a crash report -- still a user-visible error,
and no type arguments are inferred.

## Analysis

`checkAnnotationEquality` compares two `AnnotatedTypeMirror`s with `TypeHierarchy.isSubtype` in both
directions.  When both are wildcards, the stack above shows the comparison reaching one wildcard's
implicit `extends` bound (`@Tainted Object`) against the other's `super` bound (`Holder<...>`), and
`castedAsSuper` then fails to find `Holder` in `Object`.

Two conditions are needed:

  - Both sides must be lower-bounded wildcards.  `? extends` on both sides does not crash.
  - The wildcard's bound must itself be a parameterized type.  Replacing `Holder<X>` and
    `Holder<String>` with `String` reports `type.arguments.not.inferred` cleanly.

Ordinary subtyping does not hit this, because the assignment path captures the wildcard first:
`Pair<String, ? super @Tainted Holder<String>> b = a;` produces a normal `assignment` error
mentioning `capture#01`.  `checkAnnotationEquality` passes the uncaptured wildcards straight to
`isSubtype`.

Introduced by 007062239b ("Check qualifiers of equality constraints implied by an inference
variable's bounds"), which added `checkAnnotationEquality`.
